File: tp2/Questoes/Restaurantes1.java
import java.util.Scanner;

class Data{
	private int dia;
	private int mes;
	private int ano;

	public Data(int dia, int mes, int ano){
		this.dia = dia;
		this.mes = mes;
		this.ano = ano;
	}

	public static Data parseData(String d){
		Scanner sc = new Scanner(d);
		sc.useDelimiter("-");
		int ano = sc.nextInt();
		int mes = sc.nextInt();
		int dia = sc.nextInt();
		sc.close();
		return new Data(dia, mes, ano);
	}

	public String formatar(){
		return String.format("%02d/%02d/%04d", dia, mes, ano);
	}
}

class Hora{
	private int hora;
	private int min;

	public Hora(int hora, int min){
		this.hora = hora;
		this.min = min;
	}

	public static Hora parseHora(String h){
		Scanner sc = new Scanner(h);
		sc.useDelimiter(":");
		int hora = sc.nextInt();
		int min = sc.nextInt();
		sc.close();
		return new Hora(hora, min);
	}

	public String formatar(){
		return String.format("%02d:%02d", hora, min);
	}
}

class Restaurante{
	private Data dataAbertura;
	private Hora horarioAbertura;
	private Hora horarioFechamento;
	private boolean aberto;
	private int id;
	private String nome;
	private String cidade;
	private int capacidade;
	private double avaliacao;
	private String[] tiposCozinha;
	private int faixaPreco;

	public Restaurante(int id, String nome, String cidade, int capacidade, double avaliacao,
			String[] tiposCozinha, int faixaPreco,
			Hora horarioAbertura, Hora horarioFechamento,
			Data dataAbertura, boolean aberto){

		this.id = id;
		this.nome = nome;
		this.cidade = cidade;
		this.capacidade = capacidade;
		this.avaliacao = avaliacao;
		this.tiposCozinha = tiposCozinha;
		this.faixaPreco = faixaPreco;
		this.horarioAbertura = horarioAbertura;
		this.horarioFechamento = horarioFechamento;
		this.dataAbertura = dataAbertura;
		this.aberto = aberto;
	}

	public int getId(){
		return id;
	}
	public String getNome(){
		return nome;
	}
	public String getCidade(){
		return cidade;
	}
	public int getCapacidade(){
		return capacidade;
	}
	public double getAvaliacao(){
		return avaliacao;
	}
	public String[] getTiposCozinha(){
		return tiposCozinha;
	}
	public int getFaixaPreco(){
		return faixaPreco;
	}
	public Hora getHorarioAbertura(){
		return horarioAbertura;
	}
	public Hora getHorarioFechamento(){
		return horarioFechamento; 
	}
	public Data getDataAbertura(){ 
		return dataAbertura;
	}
	public boolean getAberto(){ 
		return aberto; 
	}

	public static String[] separarTipos(String tipos){
		Scanner sc = new Scanner(tipos);
		sc.useDelimiter(";");

		String[] lista = new String[5];
		int i = 0;
		while(sc.hasNext()){
			lista[i++] = sc.next();
		}

		sc.close();
		return lista;
	}

	public static int converterFaixa(String faixa){
		int num = 0;
		for(int i = 0; i < faixa.length(); i++){
			if(faixa.charAt(i) == '$'){
				num++;
			}
		}
		return num;
	}

	public static String[] separarHora(String hora){
		Scanner sc = new Scanner(hora);
		sc.useDelimiter("-");

		String abertura = sc.next();
		String fechamento = sc.next();
		sc.close();

		return new String[]{abertura, fechamento};
	}

	public static Restaurante parseRestaurante(String linha){
		Scanner sc = new Scanner(linha);
		sc.useDelimiter(",");
		sc.useLocale(java.util.Locale.US);

		int id = sc.nextInt();
		String nome = sc.next();
		String cidade = sc.next();
		int capacidade = sc.nextInt();
		double avaliacao = sc.nextDouble();

		String[] tipos = separarTipos(sc.next());
		int faixa = converterFaixa(sc.next());

		String[] horas = separarHora(sc.next());
		Hora abertura = Hora.parseHora(horas[0]);
		Hora fechamento = Hora.parseHora(horas[1]);

		Data data = Data.parseData(sc.next());

		boolean aberto = (sc.next().charAt(0) == 't');

		sc.close();

		return new Restaurante(id, nome, cidade, capacidade, avaliacao,
				tipos, faixa, abertura, fechamento, data, aberto);
	}

	public String imprimir(){
		String tipos = "";
		for(int i = 0; i < tiposCozinha.length && tiposCozinha[i] != null; i++){
			tipos += tiposCozinha[i];
			if(i < tiposCozinha.length - 1 && tiposCozinha[i+1] != null){
				tipos += ",";
			}
		}

		String faixa = "";
		for(int i = 0; i < faixaPreco; i++){
			faixa += "$";
		}

		return "[" + id + " ## " + nome + " ## " + cidade + " ## " + capacidade + " ## " + avaliacao + " ## [" + tipos + "] ## " + faixa + " ## " + horarioAbertura.formatar() + "-" + horarioFechamento.formatar() + " ## " + dataAbertura.formatar() + " ## " + aberto + "]";
	}
}

class ColecaoRestaurantes{
	private int tamanho;
	private Restaurante[] restaurantes;

	public ColecaoRestaurantes(){
		this.restaurantes = new Restaurante[502];
		this.tamanho = 0;
	}

	public int getTamanho(){ 
		return tamanho; 
	}
	public Restaurante[] getRestaurantes(){ 
		return restaurantes; 
	}

	public static ColecaoRestaurantes lerCsv(){
		ColecaoRestaurantes c = new ColecaoRestaurantes();

		try{
			Scanner sc = new Scanner(new java.io.File("/tmp/restaurantes.csv"));

			if(sc.hasNextLine()) sc.nextLine();

			while(sc.hasNextLine()){
				c.restaurantes[c.tamanho++] =
					Restaurante.parseRestaurante(sc.nextLine());
			}

			sc.close();
		}catch(Exception e){
			System.out.println("Erro ao ler");
		}

		return c;
	}
}

public class Restaurantes1 {
	public static void main(String[] args){

		Scanner sc = new Scanner(System.in);

		ColecaoRestaurantes colecao = ColecaoRestaurantes.lerCsv();
		Restaurante[] lista = colecao.getRestaurantes();

		int id = sc.nextInt();

		while(id != -1){
			for(int i = 0; i < colecao.getTamanho(); i++){
				if(lista[i].getId() == id){
					System.out.println(lista[i].imprimir());
				}
			}

			id = sc.nextInt();
		}

		sc.close();
	}
}

File: tp2/Questoes/restaurantes1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct{
    int dia;
    int mes;
    int ano;
} Data;

typedef struct{
    int hora;
    int minuto;
} Hora;

typedef struct{
    int id;
    char nome[100];
    char cidade[100];
    int capacidade;
    double avaliacao;
    char tipos_cozinha[5][50];
    int qtd_tipos;
    int faixa_preco;
    Hora horario_abertura;
    Hora horario_fechamento;
    Data data_abertura;
    int aberto;
} Restaurante;

typedef struct{
    Restaurante* restaurantes[2000];
    int tamanho;
} Colecao_restaurantes;

Data parse_data(char* s){
    Data data;
    sscanf(s, "%d-%d-%d", &data.ano, &data.mes, &data.dia);
    return data;
}

void formatar_data(Data* data, char* buffer){
    sprintf(buffer, "%02d/%02d/%04d", data->dia, data->mes, data->ano);
}

Hora parse_hora(char* s){
    Hora h;
    sscanf(s, "%d:%d", &h.hora, &h.minuto);
    return h;
}

void formatar_hora(Hora* h, char* buffer){
    sprintf(buffer, "%02d:%02d", h->hora, h->minuto);
}

int contar_faixa(char* s){
    int i = 0;
    int count = 0;

    while(s[i] != '\0'){
        if(s[i] == '$'){
            count++;
        }
        i++;
    }

    return count;
}

void separar_tipos(char* s, Restaurante* r){
    int i = 0, j = 0, k = 0;

    while(1){
        if(s[i] == ';' || s[i] == '\0'){
            r->tipos_cozinha[j][k] = '\0';
            j++;
            k = 0;

            if(s[i] == '\0'){
                break;
            }
        }else{
            r->tipos_cozinha[j][k] = s[i];
            k++;
        }
        i++;
    }

    r->qtd_tipos = j;
}

Restaurante* parse_restaurante(char* linha){
    Restaurante* r = (Restaurante*) malloc(sizeof(Restaurante));

    char tipos[200];
    char faixa[10];
    char horas[50];
    char data_str[20];
    char aberto_str[10];

    sscanf(linha, "%d,%[^,],%[^,],%d,%lf,%[^,],%[^,],%[^,],%[^,],%s",
        &r->id,
        r->nome,
        r->cidade,
        &r->capacidade,
        &r->avaliacao,
        tipos,
        faixa,
        horas,
        data_str,
        aberto_str
    );

    separar_tipos(tipos, r);
    r->faixa_preco = contar_faixa(faixa);

    char abertura[10], fechamento[10];
    sscanf(horas, "%[^-]-%s", abertura, fechamento);

    r->horario_abertura = parse_hora(abertura);
    r->horario_fechamento = parse_hora(fechamento);
    r->data_abertura = parse_data(data_str);

    if(aberto_str[0] == 't'){
        r->aberto = 1;
    }else{
        r->aberto = 0;
    }

    return r;
}

void formatar_restaurante(Restaurante* r, char* buffer){
    char data[20];
    char h_abertura[10], h_fechamento[10];

    formatar_data(&r->data_abertura, data);
    formatar_hora(&r->horario_abertura, h_abertura);
    formatar_hora(&r->horario_fechamento, h_fechamento);

    char tipos[200];
    int pos = 0;

    for(int i = 0; i < r->qtd_tipos; i++){
        int j = 0;
        while(r->tipos_cozinha[i][j] != '\0'){
            tipos[pos++] = r->tipos_cozinha[i][j];
            j++;
        }

        if(i < r->qtd_tipos - 1){
            tipos[pos++] = ',';
        }
    }
    tipos[pos] = '\0';

    char faixa[10];
    int k = 0;
    for(int i = 0; i < r->faixa_preco; i++){
        faixa[k++] = '$';
    }
    faixa[k] = '\0';

    char aberto_str[6];
    if(r->aberto == 1){
        aberto_str[0] = 't';
        aberto_str[1] = 'r';
        aberto_str[2] = 'u';
        aberto_str[3] = 'e';
        aberto_str[4] = '\0';
    }else{
        aberto_str[0] = 'f';
        aberto_str[1] = 'a';
        aberto_str[2] = 'l';
        aberto_str[3] = 's';
        aberto_str[4] = 'e';
        aberto_str[5] = '\0';
    }

    sprintf(buffer, "[%d ## %s ## %s ## %d ## %.1lf ## [%s] ## %s ## %s-%s ## %s ## %s]",
        r->id,
        r->nome,
        r->cidade,
        r->capacidade,
        r->avaliacao,
        tipos,
        faixa,
        h_abertura,
        h_fechamento,
        data,
        aberto_str
    );
}

void ler_csv_colecao(Colecao_restaurantes* c, char* path){
    FILE* file = fopen(path, "r");
    if(file == NULL){
        printf("Erro");
        return;
    }
    char linha[500];
    fgets(linha, sizeof(linha), file);

    while(fgets(linha, sizeof(linha), file)){
        int i = 0;
        while(linha[i] != '\0'){
            if(linha[i] == '\n'){
                linha[i] = '\0';
            }
            i++;
        }

        c->restaurantes[c->tamanho++] = parse_restaurante(linha);
    }

    fclose(file);
}

Colecao_restaurantes* ler_csv(){
    Colecao_restaurantes* c = (Colecao_restaurantes*) malloc(sizeof(Colecao_restaurantes));
    c->tamanho = 0;
    ler_csv_colecao(c, "/tmp/restaurantes.csv");
    return c;
}

int main(){
    Colecao_restaurantes* colecao = ler_csv();
    int id;
    scanf("%d", &id);

    while(id != -1){
        for(int i = 0; i < colecao->tamanho; i++){
            if(colecao->restaurantes[i]->id == id){
                char buffer[500];
                formatar_restaurante(colecao->restaurantes[i], buffer);
                printf("%s\n", buffer);
            }
        }

        scanf("%d", &id);
    }

    return 0;
}
